[로또 - 웹, DB] 무민 미션 제출합니다.

green4469 opened the pull request:
많이 부족한걸 알지만 일단 제출해봅니다 ㅠㅠ.

DAO 테스트 작성이 너무 어렵네요. DB 상태를 원상복귀 시키면서 각 기능을 테스트하고자 하는데 어떻게 해야할 지 잘 모르겠습니다.

잘못된 부분 많은 지적 부탁드립니다!

green4469 committed:
feat: web ui

green4469 committed:
add ResultDAO and database connector

green4469 committed:
add DAOs and store data in DB

green4469 committed:
show all rounds read from db

green4469 committed:
develop RESTful server, vue works with sparkjava

phs1116 commented on resources/tableCreation.sql:
```diff
@@ -0,0 +1,76 @@
+    PRIMARY KEY (id)
+);
+
+ALTER TABLE winningLotto;
```
ALTER 를 다시 치는 이유에 대해서 알 수 있을까요?

green4469 replied:
AqueryTool 사이트(https://aquerytool.com/)에서 ERD를 그리고, SQL 자동 생성기능을 사용했습니다.

ALTER를 다시 치지 않도록 코드를 수정했습니다. 확인 부탁드려요.

phs1116 commented on src/main/java/lotto/WebUILottoApplication.java:
```diff
@@ -1,19 +1,109 @@
+
+        get("/result", (req, res) -> {
+            // get data from section
+            WinningLotto winningLotto = req.session().attribute("winningLotto");
```
각 코드들이 너무 추상화 되어있지 않고 세부 구현이 드러나있어서 의도를 파악하기 쉽지 않습니다.
DAO외에 서비스 레이어나 컨트롤러 레이어를 두고 추상화해보면 어떨까요?

green4469 replied:
MVC 로만 하다가, 이번에 주신 피드백을 반영하며 처음으로 레이어드 아키텍처 개념을 공부했습니다.

컨트롤러 아래에 서비스 레이어를 추가하였습니다. 이로 인해 아래와 같은 변화가 발생했습니다.
- 이제 컨트롤러 레이어에서 직접 도메인 객체를 생성하거나 그의 메서드를 호출하지 않습니다. 
- 컨트롤러는 도메인 모델이 어떻게 구현됐는지 알지 못하고 서비스 객체들에만 의존합니다. 
- 컨트롤러는 더이상 DAO를 호출하지 않습니다.
- 컨트롤러 레이어와 서비스 레이어는 자바 기본형 혹은 DTO로 데이터를 주고받습니다.
- DTO를 도메인 모델에서 서비스 레이어로 옮겼습니다.

로또 웹 프로그램을 Presentation Layer (Controller) - Service Layer (Domain Model) - Persistence Layer (DAOs) 로 나누려 노력했습니다. 잘못된 점 많은 지적 부탁드립니다.

항상 감사합니다.

phs1116 commented on src/test/java/lotto/database/ResultDAOTest.java:
```diff
@@ -0,0 +1,57 @@
+import java.util.HashMap;
+import java.util.Map;
+
+public class ResultDAOTest {
```
DAO Test에 대해서는 방법은 두가지가 있을 것 같네요.
1. 테스트 종료 후에 롤백한다.
2.테스트 시에만 메모리 DB를 사용하도록 한다.(H2 등)

green4469 replied:
round를 생성하고, round를 Foreign Key 로 갖는 테이블의 레코드를 생성하는 방식으로 테스트했습니다. 확인 부탁드립니다.

phs1116 commented on src/main/java/lotto/view/WebOutputView.java:
```diff
@@ -0,0 +1,45 @@
+        return lottos.getLottos().stream().map(lotto -> outputLotto(lotto)).collect(Collectors.toList());
+    }
+
+    public static List<String> outputResult(WinningStatistics winningStatistics) {
```
MVC 관점에서 View에 의존적인 내용은 View단에서 다 알아서 하는게 좋지 않을까싶네요. :)

green4469 replied:
객체 -> 문자열표현 으로 변환하는 로직을 DTO 안으로 옮겼습니다.

phs1116 commented on src/main/java/lotto/view/WebInputView.java:
```diff
@@ -0,0 +1,50 @@
+import java.util.regex.Pattern;
+import java.util.stream.Collectors;
+
+public class WebInputView {
```
WebInputView가 전역적으로 활용될 위험이 있는 것 같습니다.
사용되는 부분은 각 요청(컨트롤러)에만 국한되는데 말이죠. :)

green4469 replied:
WebInputView를 사용하지 않도록 수정했습니다.

phs1116 commented on src/main/java/lotto/database/WinningLottoDAO.java:
```diff
@@ -0,0 +1,52 @@
+
+        List<Integer> numbers = winningLotto.getLotto().getLottoNumbers().stream().map(n -> n.getNumber()).collect(Collectors.toList());
+        for (int i = 0; i < numbers.size(); i++) {
+            pstmt.setString(2+i, String.valueOf(numbers.get(i)));
```
`2+i` 오토포맷팅이 안먹힌것 같네요. :)

green4469 replied:
수정했습니다!

phs1116 reviewed (changes requested):
안녕하세요! 리뷰어 화투입니다.
기존 도메인 코드를 건드리지 않고 웹을 잘 적용하신 것 같아요. 👍 
다만 코드에 대한 구현이 너무 드러나있어서 약간이나마 추상화해보면 좋을 것 같아 피드백 남겨드렸어요.
참고 부탁드릴게요 :) 궁금하신 점 있으시면 DM 주세요!

green4469 committed:
refactor: add service layer

Introduce 3-tier layered architecture. (Presentation Layer - Service Layer
- Persistence Layer)

green4469 committed:
refactor: add tests for DAOs

phs1116 commented on src/main/java/lotto/WebUILottoApplication.java:
```diff
@@ -1,19 +1,73 @@
         });
+
+        post("/home", (req, res) -> {
+            int lottoBuyingMoney = LottoBuyingMoneyService.validateLottoBuyingMoney(req.queryParams("lottoBuyingMoney"));
```
레이어드 아키텍쳐를 공부하신 김에 해당 구현부들은 Presentation Layer의 Controller 클래스로 분리할 수 있을 것 같네요. :)

green4469 replied:
Controller 클래스를 만들어서 로직을 분리해봤습니다. 확인 부탁드립니다.

phs1116 commented on src/main/java/lotto/persistence/Connector.java:
```diff
@@ -0,0 +1,44 @@
+                    + database + "?serverTimezone=UTC&useSSL=false", userName, password);
+            System.out.println("데이터베이스에 연결되었습니다.");
+        } catch (SQLException e) {
+            System.err.println("데이터베이스 연결 오류: " + e.getMessage());
```
해당 구현처럼 예외를 잡을시에는 Dao에서 NPE로 예외가 발생할 것으로 보이네요. :)
NPE로 발생시에는 상세 stackTrace나 로그내용을 찾지 않은 이상 원인을 바로 찾기 쉽지 않을 것 같은데요.
커넥션 생성에 실패시에는 별도의 예외를 던지는건 어떨까요?

green4469 replied:
FailedDBConnectionException 예외를 만들고 이를 던지도록 수정했습니다.

phs1116 commented on src/main/java/lotto/service/LottoService.java:
```diff
@@ -0,0 +1,51 @@
+        resultDao.addResult(thisRoundId, winStat.getStatistics());
+        winningLottoDao.addWinningLotto(thisRoundId, winLotto);
+        lottoDao.addLotto(thisRoundId, lottos);
+        Connector.closeConnection(con);
```
try-with-resrouces로 커넥션을 닫을 수 있을 것 같네요!
켜넥션에 대해서도 Service가 직접 알아야할지도 고민해보시면 좋을 것 같습니다. :)
Connection를 Service가 알고있다는 말은 Dao의 구현에 대해서 자세히 알고있는걸로 보여서요!

green4469 replied:
try-with-resources 라는 게 존재하는지 처음 알았습니다. 반영하여 코드를 수정했습니다.

green4469 replied:
말씀하신대로 서비스에서 Connection을 직접 생성하는것은 DAO의 캡슐화에 좋지 않다고 생각합니다. Connection을 서비스에서 생성하지 않고, DAO가 직접 생성해서 사용하도록 코드를 수정했습니다.

phs1116 commented on src/main/java/lotto/persistence/RoundDAO.java:
```diff
@@ -0,0 +1,77 @@
+import java.util.ArrayList;
+import java.util.List;
+
+public class RoundDAO {
```
각 서비스에서 DAO의 구현체를 직접 참조하지 않고, 인터페이스를 만들고 이를 통해서 참조하도록 해보면 어떨까요?
그리고 인터페이스를 작성했을 경우에 얻는 잇점은 무엇일까요?

green4469 replied:
데이터베이스를 여러 개 사용할 경우에 이점이 있을 것 같습니다.

MySQL과 Oracle 데이터베이스를 모두 사용한다면 각각에 대한 DAO 구현체가 필요할 것입니다. (데이터베이스의 SQL 문법이 미세하게 다르다던지, 테이블 네이밍이 다르게 되어있다던지) 

DAO 인터페이스를 만들고 서비스가 이를 통해서 참조한다면 구현체를 생성하는 코드만 바꿔주면 되기때문에 서비스가 DAO가 어떤 DB와 커플링되어 있는지 신경쓰지 않아도 될 것 같습니다.

phs1116 commented on src/main/java/lotto/persistence/LottoDAO.java:
```diff
@@ -0,0 +1,62 @@
+        this.con = con;
+    }
+
+    public void addLotto(int roundId, Lottos lottos) throws SQLException {
```
현재 SQLException를 잡지 않고 직접 던지고 있는데요, 요새 많이 안티패턴으로 자리잡고 있는 방식중 하나입니다.

아래 블로그 내용을 공부하고 예외를 핸들링해보면 어떨까요?
http://wonwoo.ml/index.php/post/878

green4469 replied:
DAO 에서 SQLException을 던지는 경우, 이를 호출한 서비스 계층에서 예외를 해결할 수가 없기 때문에 폭탄돌리기가 시작되고, SQLException 은 반드시 catch로 잡아주어야하는 예외이기 때문에 throws, throws 가 반복되는 문제가 발생함을 이해했습니다.

위의 문제를 해결하기위해 SQLException을 잡고, 이를 RuntimeException으로 포장해서 던지도록 코드를 수정했습니다.

phs1116 commented on src/main/java/lotto/persistence/LottoDAO.java:
```diff
@@ -0,0 +1,62 @@
+        String query = "SELECT * FROM lotto WHERE ro_id=?";
+        PreparedStatement pstmt = con.prepareStatement(query);
+        pstmt.setString(1, String.valueOf(roundId));
+        ResultSet rs = pstmt.executeQuery();
```
ResultSet에 대해서도 꼭 close 해주세요!

green4469 replied:
수정했습니다.

phs1116 commented on src/main/java/lotto/persistence/RoundDAO.java:
```diff
@@ -0,0 +1,77 @@
+    }
+
+    public void addRound(int prize, double interestRate) throws SQLException {
+        String query = "INSERT INTO round (prize, interest_rate) VALUES(?, ?)";
```
DAO 같은 패턴으로 작성된 코드가 상당히 많은데요.
PreparedStatement 생성, 쿼리 실행, 매핑 등
이를 별도 템플릿 클래스를 만들어 분리해보면 어떨까요?

green4469 replied:
쿼리를 날릴때마다 반복되는 작업을 JDBCTemplate 클래스로 빼내고 이를 DAO에서 사용하도록 코드를 수정했습니다.

phs1116 commented on src/main/java/lotto/service/LottoService.java:
```diff
@@ -0,0 +1,51 @@
+import static lotto.service.ServiceUtils.parseInts;
+
+public class LottoService {
+    public static void saveResult(int lottoBuyingMoney,
```
static 메서드가 아닌 인스턴스로 작성해보면 어떨까요? 그 경우 얻을수 있는 잇점이 무엇이 있을지도 고민해보시면 좋을 것 같네요. :)

green4469 replied:
서비스 내에서 상태관리를 할 수 있을 것 같습니다. static 메서드로 서비스를 사용할 경우 불필요하게 서비스 로직이 집중돼 비대해지는 것을 느꼈습니다.

phs1116 replied:
서비스에서는 일반적으로 상태 관리를 가지지 않도록 하는게 정석입니다. :) 
상태랑은 다르지만 static 메서드가 아닌 경우에 대해서는 LottoService는 자신의 의존하는 객체를 필드로 가질 수 있을 것이고 (DAO 등), 생성자를 통해 이들을 동적으로 주입을 받을 수 있겠네요!
(스프링 DI 컨테이너, 의존성 주입,  IoC등을 참고해보시면 좋을 것 같습니다. :))
그리고 static이 아닌 인스턴스로 관리할 경우에는 상속을 통한 Mock으로 사용 가능하여 유연한 테스트도 가능합니다.

phs1116 reviewed (changes requested):
안녕하세요! 리뷰어 화투입니다.
서비스 클래스 분리 👍 
다만 예외 핸들링에 대한 부분과 추상화 부분에 대해서 피드백 몇가지 추가했습니다.
참고 부탁드리고 궁금하신 점 있으시면 언제든지 연락주세요.:)

green4469 committed:
refactor: abstract controller into multiple controller classes

green4469 committed:
refactor: Throw custom exceptions when DB Connection fails

green4469 committed:
refactor: use try-with-resources to close connection, close ResultSet in DAOs

green4469 committed:
refactor: Abstract DAOs into interfaces, deal with SQLExceptions thrown by JDBC Driver in DAOs.

green4469 committed:
refactor: ResultSet close error fix, remove magic numbers

green4469 committed:
refactor: Introduce JDBCTemplate, which abstracts JDBC related operations.

green4469 committed:
refactor: Use JDBCTemplate in ResultDAO, WinningLottoDAO and RoundDAO remained

green4469 committed:
refactor: Use JDBCTemplate in all DAOs

phs1116 commented on src/main/java/lotto/persistence/Connector.java:
```diff
@@ -0,0 +1,48 @@
+            Class.forName("com.mysql.cj.jdbc.Driver");
+        } catch (ClassNotFoundException e) {
+            System.err.println((" !! JDBC Driver load 오류: " + e.getMessage()));
+            e.printStackTrace();
```
e.printStackTrace();는 일반적으로 안티패턴으로 간주되니 빼주시면 좋을 것 같아요. :)

phs1116 reviewed (approved):
안녕하세요! 리뷰어 화투입니다.
피드백 반영 및 정리 💯 
이번 단계 머지할게요.
다만 정리한 사항에 대한 추가 코멘트랑 간단한 피드백 한개 추가했는데,
나중에 진행하면서 참고 부탁드리겠습니다. :)
개발하시느라 수고 많으셨습니다!
